monitor/loadgen/xcall.go

```diff
@@ -0,0 +1,131 @@
+	"time"
+
+	"github.com/omni-network/omni/contracts/bindings"
+	"github.com/omni-network/omni/lib/contracts"
+	"github.com/omni-network/omni/lib/errors"
+	"github.com/omni-network/omni/lib/ethclient/ethbackend"
+	"github.com/omni-network/omni/lib/log"
+	"github.com/omni-network/omni/lib/netconf"
+	"github.com/omni-network/omni/lib/xchain"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+const (
+	dead = "0x000000000000000000000000000000000000dead"
+)
+
+type XCallConfig struct {
+	NetworkID   netconf.ID
+	XCallerAddr common.Address
+	Period      time.Duration
+	Backends    ethbackend.Backends
+	Chains      []netconf.Chain
+}
+
+func xCallForever(ctx context.Context, cfg *XCallConfig) {
+	log.Info(ctx, "Starting periodic xcalls", "period", cfg.Period)
+
+	nextPeriod := func() time.Duration {
+		jitter := time.Duration(float64(cfg.Period) * rand.Float64() * loadgenJitter) //nolint:gosec // Weak random ok for load tests.
+		return cfg.Period + jitter
+	}
+
+	timer := time.NewTimer(nextPeriod())
+	defer timer.Stop()
+
+	// returns random chain pair
+	getChainPair := func() (netconf.Chain, netconf.Chain) {
+		i := rand.IntN(len(cfg.Chains)) //nolint:gosec // Weak random ok for load tests.
+		j := rand.IntN(len(cfg.Chains)) //nolint:gosec // Weak random ok for load tests.
+		for i == j {
+			j = rand.IntN(len(cfg.Chains)) //nolint:gosec // Weak random ok for load tests.
+		}
+
+		return cfg.Chains[i], cfg.Chains[j]
+	}
+
+	// tick immediately
+	if err := xCall(ctx, cfg, getChainPair); err != nil {
+		log.Warn(ctx, "Failed to xcall (will retry)", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			if err := xCall(ctx, cfg, getChainPair); err != nil {
+				log.Warn(ctx, "Failed to xcall (will retry)", err)
+			}
+			timer.Reset(nextPeriod())
+		}
+	}
+}
+
+func xCall(ctx context.Context, cfg *XCallConfig, getChainPair func() (from netconf.Chain, to netconf.Chain)) error {
+	fromChain, dstChain := getChainPair()
+	backend, err := cfg.Backends.Backend(fromChain.ID)
+	if err != nil {
+		return err
+	}
+
+	fromPortal, err := getPortal(ctx, cfg.NetworkID, fromChain.ID, cfg.Backends)
+	if err != nil {
+		return err
+	}
+
+	var data []byte
+	to := common.HexToAddress(dead)
+	gasLimit := uint64(100_000)
+	fee, err := fromPortal.FeeFor(&bind.CallOpts{}, dstChain.ID, data, gasLimit)
+	if err != nil {
+		return errors.Wrap(err, "feeFor",
+			"src_chain", fromChain.ID,
+			"dst_chain_id", dstChain.ID,
+		)
+	}
+
+	txOpts, err := backend.BindOpts(ctx, cfg.XCallerAddr)
+	if err != nil {
+		return errors.Wrap(err, "bindOpts")
+	}
+
+	txOpts.Value = fee
+	tx, err := fromPortal.Xcall(txOpts, dstChain.ID, uint8(xchain.ConfLatest), to, data, gasLimit)
+	if err != nil {
+		return errors.Wrap(err, "xcall",
+			"src_chain", fromChain.ID,
+			"dst_chain_id", dstChain.ID,
+		)
+	}
+	log.Debug(ctx, fmt.Sprintf("xcall made %s -> %s %s", fromChain.Name, dstChain.Name, tx.Hash()))
```

Comment:
avoid %s or %v in logging, use structured log attributes rather

Reply:
https://github.com/omni-network/omni/pull/2723/commits/21f8d15443640dce3209ee379fe9ae5fbd3d32e6